main.cpp

```diff
@@ -1,75 +1,96 @@
-#include <clang/Frontend/CompilerInstance.h>
-
-using namespace clang;
-using namespace clang::ast_matchers;
-using namespace clang::tooling;
-
-class CastCallBack : public MatchFinder::MatchCallback {
-public:
-    CastCallBack(Rewriter& rewriter) {
-        // Your code goes here
-    };
-
-    void run(const MatchFinder::MatchResult &Result) override {
-        // Your code goes here
-    }
-};
-
-class MyASTConsumer : public ASTConsumer {
-public:
-    MyASTConsumer(Rewriter &rewriter) : callback_(rewriter) {
-        matcher_.addMatcher(
-                cStyleCastExpr(unless(isExpansionInSystemHeader())).bind("cast"), &callback_);
-    }
-
-    void HandleTranslationUnit(ASTContext &Context) override {
-        matcher_.matchAST(Context);
-    }
-
-private:
-    CastCallBack callback_;
-    MatchFinder matcher_;
-};
-
-class CStyleCheckerFrontendAction : public ASTFrontendAction {
-public:
-    CStyleCheckerFrontendAction() = default;
-    
-    void EndSourceFileAction() override {
-        rewriter_.getEditBuffer(rewriter_.getSourceMgr().getMainFileID())
-            .write(llvm::outs());
-    }
-
-    std::unique_ptr<ASTConsumer> CreateASTConsumer(CompilerInstance &CI, StringRef /* file */) override {
-        rewriter_.setSourceMgr(CI.getSourceManager(), CI.getLangOpts());
-        return std::make_unique<MyASTConsumer>(rewriter_);
-    }
-
-private:
-    Rewriter rewriter_;
-};
-
-static llvm::cl::OptionCategory CastMatcherCategory("cast-matcher options");
-
-int main(int argc, const char **argv) {
-    // Ubuntu 20
-    auto Parser = llvm::ExitOnError()(CommonOptionsParser::create(argc, argv, CastMatcherCategory));
-    // For Ubuntu 18, use the old Clang API to construct the CommonOptionsParser:
-    // CommonOptionsParser Parser(argc, argv, CastMatcherCategory);
-
-    ClangTool Tool(Parser.getCompilations(), Parser.getSourcePathList());
-    return Tool.run(newFrontendActionFactory<CStyleCheckerFrontendAction>().get());
-}
+#include <iostream>
+#include <sstream>
+
+#include <llvm/Support/CommandLine.h>
+#include <llvm/Support/Error.h>
+
+#include <clang/ASTMatchers/ASTMatchFinder.h>
+#include <clang/ASTMatchers/ASTMatchers.h>
+#include <clang/Basic/Diagnostic.h>
+#include <clang/Frontend/CompilerInstance.h>
+#include <clang/Frontend/FrontendActions.h>
+#include <clang/Rewrite/Core/Rewriter.h>
+#include <clang/Tooling/CommonOptionsParser.h>
+#include <clang/Tooling/Tooling.h>
+
+using namespace clang;
+using namespace clang::ast_matchers;
+using namespace clang::tooling;
+
+class CastCallBack : public MatchFinder::MatchCallback {
+
+private:
+  Rewriter& _rewriter;
+
+public:
+  CastCallBack(Rewriter &rewriter) : _rewriter(rewriter){};
+  void run(const MatchFinder::MatchResult &Result) override {
+
+    const auto *castExpr = Result.Nodes.getNodeAs<CStyleCastExpr>("cast");
+    
+    const auto range = CharSourceRange::getCharRange(
+        castExpr->getLParenLoc(),
+        castExpr->getSubExprAsWritten()->getBeginLoc());
```

Comment:
Please describe the difference between `getSubExprAsWritten()->getBeginLoc()` and `getRParenLoc().getLocWithOffset(1)`.

Reply:
getSubExprAsWritten()->getBeginLoc() points to the beginning of the variable that we're trying to cast
getRParenLoc().getLocWithOffset(1) points to the character that's right after the closing parenthesis
Which means if we use it, we're going to have problems with spaces after the closing parenthesis